cmd/update.go

```diff
@@ -0,0 +1,496 @@
+}
+
+func findAsset(rel *ghRelease) (name, url string, err error) {
+	return findAssetFor(rel, runtime.GOOS, runtime.GOARCH)
+}
+
+func findSha256For(rel *ghRelease, binName string) (url string, ok bool) {
+	want := binName + ".sha256"
+	for _, a := range rel.Assets {
+		if a.Name == want {
+			return a.BrowserDownloadURL, true
+		}
+	}
+	return "", false
+}
+
+// Download with progress
+
+type progressWriter struct {
+	written    int64
+	lastPrint  time.Time
+	contentLen int64
+}
+
+func (p *progressWriter) Write(b []byte) (int, error) {
+	n := len(b)
+	p.written += int64(n)
+	now := time.Now()
+
+	if now.Sub(p.lastPrint) >= progressPrintFrequency || p.written == p.contentLen {
+		if p.contentLen > 0 {
+			percent := float64(p.written) / float64(p.contentLen) * 100
+			fmt.Fprintf(os.Stderr, "\rDownloading... %5.1f%%", percent)
+			if p.written == p.contentLen {
+				fmt.Fprintln(os.Stderr)
+			}
+		} else {
+			fmt.Fprintf(os.Stderr, "\rDownloading... %d KiB", p.written/1024)
+		}
+		p.lastPrint = now
+	}
+	return n, nil
+}
+
+func download(ctx context.Context, url string) (string, error) {
+	if !strings.HasPrefix(url, "https://") {
+		return "", errors.New("refusing to download from non‑HTTPS URL")
+	}
+
+	req, _ := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	res, err := httpClient.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("download failed: %s", res.Status)
+	}
+
+	tmp, err := os.CreateTemp("", "oasis-update-*")
+	if err != nil {
+		return "", err
+	}
+	defer tmp.Close()
+
+	pw := &progressWriter{contentLen: res.ContentLength}
+	_, err = io.Copy(tmp, io.TeeReader(res.Body, pw))
+	if err != nil {
+		return "", err
+	}
+	return tmp.Name(), nil
+}
+
+// Install helpers
+
+func replaceExecutable(dest string, r io.Reader, perm os.FileMode) error {
+	dir := filepath.Dir(dest)
+
+	tmp, err := os.CreateTemp(dir, ".oasis-new-*")
+	if err != nil {
+		return err
+	}
+	tmpPath := tmp.Name()
+
+	// By default clean up both the file and the descriptor on return.
+	cleanup := true
+	defer func() {
+		if cleanup {
+			_ = tmp.Close()
+			_ = os.Remove(tmpPath)
+		}
+	}()
+
+	if _, err = io.Copy(tmp, r); err != nil {
+		return err
+	}
+
+	if perm == 0 {
+		perm = 0o755
+	}
+	if err = tmp.Chmod(perm); err != nil && runtime.GOOS != osWindows {
+		return err
+	}
+
+	// Close before attempting to rename.
+	if err = tmp.Close(); err != nil {
+		return err
+	}
+
+	// POSIX: atomic rename overwrites the original executable.
+	if err = os.Rename(tmpPath, dest); err == nil {
+		cleanup = false
+		return nil
+	}
+
+	// Windows: target executable is locked – stage as .new for next run.
+	if runtime.GOOS == osWindows {
+		pending := dest + ".new"
+		if err2 := os.Rename(tmpPath, pending); err2 != nil {
+			return fmt.Errorf("rename failed (%v / %v)", err, err2)
+		}
+		cleanup = false
+		fmt.Println("Update staged. Restart oasis CLI to finish installation.")
```

Comment:
How does this work exactly? Because if you restart the CLI, then the binary is locked again.